bashbrew/go/src/bashbrew/cmd-build.go

```diff
@@ -90,6 +92,8 @@ func cmdBuild(c *cli.Context) error {
 					}
 					defer archive.Close()
 
+					// TODO use "meta.StageNames" to do "docker build --target" so we can tag intermediate stages too for cache (streaming "git archive" directly to "docker build" makes that a little hard to accomplish without re-streaming)
```

Comment:
IMO this catch is worth pointing out explicitly here too -- we don't ever have a "local" copy of the context, and instead stream `git archive` directly to `docker build`, so even if `docker build --target=N` were possible to do for a numbered stage, we'd still have an optimization problem implementing it (and I'd love to at least tag named stages that way, but the `git archive | docker build` optimization makes that a little harder to swallow).

The practical implication of this is that build cache for these untagged stages will still be considered "ripe" by our tooling and thus we will end up spending extra time rebuilding things after each cleaning of "ripe" images on our build servers.

For this reason, I think that for the official images, we should plan to still discourage multi-stage image use unless the gains are clear and/or the final artifacts are reproducible (and thus can avoid unnecessary image updates), and even then encourage use of explicitly named stages so that we can hopefully tag them in the future to help avoid this issue.